WebView handler improvements for advanced requests and Edge handler implementation

TcT2k opened the pull request:
The current `wxWebViewHandler` does only allow simple synchronous GET requests which is fine for many applications. To allow local REST like requests asynchronous POST requests, should be able to be processed.

I've added a new method `wxWebViewHandler::StartRequest()` which allows for more advanced request processing while maintaining backward compatibility through a default implementation that calls `wxWebViewHandler::GetFile()`.

A response can be send via the response object. The response does not have to be finished from this method and it's possible to be finished asynchronously via `wxWebViewHandlerResponse::Finish()`.

The following pseudo code demonstrates a typical implementation:
```C++
void MyAdvancedWebViewHandler::StartRequest(const wxWebViewHandlerRequest& request,
                      wxSharedPtr<wxWebViewHandlerResponse> response) wxOVERRIDE
{
    // Set common headers allowing acces from XMLHTTPRequests()
    response->SetHeader("Access-Control-Allow-Origin", "*");
    response->SetHeader("Access-Control-Allow-Headers", "*");

    // Handle options request
    if (request.GetMethod().IsSameAs("options", false))
    {
        response->Finish("");
        return;
    }

    // Check the post data type
    if (!request.GetHeader("Content-Type").StartsWith("application/json"))
    {
        response->FinishWithError();
        return;
    }

    // Process input data
    wxString postData = request.GetDataString();

    ...

    // Send result
    response->SetContentType("application/json");
    response->Finish("{ result: true }");
}
```

This is currently implemented for **WKWebView** and **Edge** backends. It should be possible for **webkit2** but I will probably not implement this in the first go.

The Edge backend currently does not support custom schemes. The handler implementation relies on "virtual hosts" which make handlers available at something like https://wxfs.wxsite . `LoadURL()` has handling to redirect a call to wxfs:///some.zip#protocol=zip/index.html to https://wxfs.wxsite///some.zip#protocol=zip/index.html . This allows for easy cross platform code in many cases, but obviously doesn't handle cases where the content requests custom schemes. 

In future versions custom schemes will be available via WebView2 but this is an acceptable compromise for now and can be interesting for some cases in the future, where virtual http(s) hosting of local assets might be wished.
The virtual host name defaults to the handlers name + ".wxsite" but can be customized via `wxWebViewHandler::SetVirtualHost()`.

As always any **feedback** is welcome on interface or implementation details. 
The edge backend support can be tested via the _webviewsample_.

ToDo:
- [x] Add macOS implementation
- [x] Add Edge implementation
- [x] Complete documentation
- [x] Integrate advanced webviewhandler into sample

TcT2k committed:
Add COM stream adapter

This allows easy usage of IStream based COM APIs

TcT2k committed:
Edge wxWebViewHandler implementation

Currently custom protocols are not supported. Implement using
virtual hosts and emulate custom protocol loading in LoadURI().
While this is not a exact match it will at least work without
code changes in many common cases.

TcT2k committed:
Add wxWebViewHandler::StartRequest()

The new method an associated classes allow for a more detailed
handling in wxWebViewHandler instances.

Most notibly post data and headers from the request can be processsed.
The response can be send asynchronously.

TcT2k committed:
Add advanced handler to webview sample

Demonstrate how to implement and use wxWebViewHandler::StartRequest()

TcT2k committed:
Document wxWebView::StartRequest()

Document the new method on associated classes.

NewPagodi commented:
This is great.  I wrote something similar a while back for pulling files from an sqlite database to serve to a webview a while back.  I had it working on  windows with the internet explorer backend and gtk with webkit2gtk.  But I ultimately gave up because the performance was so dreadful on windows due to the ancient backend.  I can't seem to find the files now, but I remember it being surprisingly easy to get it working with webkit2gtk.

vadz commented on include/wx/msw/private/comstream.h:
```diff
@@ -0,0 +1,179 @@
+ADD_IID(Stream)
+END_IID_TABLE;
+
+IMPLEMENT_IUNKNOWN_METHODS(wxCOMBaseStreamAdapter);
```
This is going to result in link errors if this header is included more than once, so it'd be better to move it into some existing file if possible.

vadz commented on interface/wx/webview.h:
```diff
@@ -341,7 +569,8 @@ class wxWebViewHandler
     It is available for Windows 7 and newer.
 
-    This backend does not support custom schemes and virtual file systems.
+    This backend does not support custom schemes. When using handlers see
+    wxWebViewHandler::SetVirtualHost() for more details on how to use handlers.
```
Minor, but "when using handlers ... on how to use handlers" seems a bit weird.

vadz commented on src/common/webview.cpp:
```diff
@@ -52,6 +54,92 @@ wxDEFINE_EVENT( wxEVT_WEBVIEW_FULLSCREEN_CHANGED, wxWebViewEvent);
+    size_t length = data->GetLength();
+    wxMemoryBuffer buffer;
+    data->ReadAll(buffer.GetWriteBuf(length), length);
+    wxString dataStr((const char*) buffer.GetData(), conv, length);
```
Just to show a good example:
```suggestion
    wxString dataStr(static_cast<const char*>(buffer.GetData()), conv, length);
```

vadz commented on interface/wx/webview.h:
```diff
@@ -161,6 +161,155 @@ enum wxWebViewIE_EmulationLevel
+    virtual wxInputStream* GetData() const = 0;
+
+    /**
+        @return The body data of the request as a string.
```
Might be worth to mention that the returned string is empty if the encoding is wrong?

vadz commented on interface/wx/webview.h:
```diff
@@ -161,6 +161,155 @@ enum wxWebViewIE_EmulationLevel
+            String value of the header
+    */
+    virtual void SetHeader(const wxString& name, const wxString& value) = 0;
+    /**
```
Extremely minor:
```suggestion

    /**
```

vadz commented on interface/wx/webview.h:
```diff
@@ -297,6 +450,81 @@ class wxWebViewHandler
+    /**
+        When using the edge backend handler urls are https urls with a
+        virtual host. As default @c name.wxsite is used as the virtual hostname.
+        If you customize this host use a non existing site (ideally a reserved
```
Minor too but helps readability a bit IMHO
```suggestion
        If you customize this host, use a non existing site (ideally a reserved
```

vadz commented on interface/wx/webview.h:
```diff
@@ -297,6 +450,81 @@ class wxWebViewHandler
+            response->SetHeader("Access-Control-Allow-Origin", "*");
+            response->SetHeader("Access-Control-Allow-Headers", "*");
+
+            // Handle options request
```
Do we _have_ to do this? I.e. do we get `OPTIONS` request even if we don't generate it ourselves? Or is for demonstration purposes only (but then why does the same do it too)?

TcT2k replied:
It depends on the context, but a `XMLHTTPRequest` generates (regardless of the requests HTTP method) a `OPTIONS` request from the browser to check cross site security.

vadz reviewed (approved):
Thanks, looks excellent, as usual, but just a question: is it feasible to implement this with GTK webkit later or does it not support this functionality at all?

TcT2k committed:
Apply suggestions from code review

Co-authored-by: VZ <vz-github@zeitlins.org>

TcT2k committed:
Additional wxWebView documentation tweaks

TcT2k committed:
Reimplement wxCOMStreamAdapterBase COM interface

Add inline implementation of IUnkown methods to avoid link
errors if comstream.h would be included multiple times.

TcT2k commented:
> This is going to result in link errors if this header is included more than once, so it'd be better to move it into some existing file if possible.

I've reimplemented the few `IUnknown` methods required inline as the `comimpl.h` macros don't support exported classes and this was an easier fix for now.

Additionally I've integrated your documentation suggestions.

> Thanks, looks excellent, as usual, but just a question: is it feasible to implement this with GTK webkit later or does it not support this functionality at all?

Thanks, from the API documentation I looked at, everything except access to the requests HTTP body (e.g. POST data) should be possible with GTK webkit. As stated I won't implement it at this time. I might add it in the future, if I find the time and nobody beats me too it.

vadz commented:
Thanks again and it's definitely not a problem to not add webkit implementation right now, I just wanted to know if it were possible in the future, thanks for confirming this.

I'll merge it when the CI builds finish.
